### PR TITLE
ICDS Dashboard: Service Delivery Dashboard Mobile UI

### PR DESCRIPTION
**Pull request by saiprakash-v**
Trello: https://trello.com/c/8iBlwmIN
##### SUMMARY

Adding SDD mobile UI. Cards in the UI is a seperate component "sortable-kpi".
To make the ajax request for data (10 entries at a time, order and column of sorting), I have added params explicitly (which in case of web dashboard "angular-datatable" library takes care). After successful request, I have generated necessary data for sortable-kpi to be displayed from the response.

Also when user reaches bottom of the page and there are entries that are yet to be fetched, triggering request to get next set of data.
##### FEATURE FLAG

Only featured users can see SDD in nav menu
##### PRODUCT DESCRIPTION

SDD in nav Menu:
Screenshot from 2020-03-03 18-52-43
expanded and collapsed cards:
Screenshot from 2020-03-03 18-55-23
Sort Popup: (Clickable headings to sort data)
Screenshot from 2020-03-03 18-56-05
Alert showing SDD reports availability:
Screenshot from 2020-03-03 18-56-47

### CODE REVIEW

File: custom/icds_reports/static/icds_reports/js/spec/month-filter.directive.spec.js [hunk 1]
```diff
@@ -15,6 +15,7 @@ describe('Month Filter Controller', function () {
     beforeEach(module('icdsApp', function ($provide) {
         $provide.constant("userLocationId", null);
         $provide.constant("isAlertActive", false);
+        $provide.constant("isMobile", false);
```
Comment (czue) [inline custom/icds_reports/static/icds_reports/js/spec/month-filter.directive.spec.js]: nit: you could consider migrating this to `utils.provideDefaultConstants`

File: custom/icds_reports/static/icds_reports/mobile/css/app.less [hunk 2]
```diff
@@ -326,3 +329,40 @@
 .alert-maintenance .close {
   margin-left: 3px;
 }
+
+.text-underline {
+  text-decoration: underline;
+}
+
+.sort-popup {
+  width: 280px;
```
Comment (czue) [inline custom/icds_reports/static/icds_reports/mobile/css/app.less]: nit: should this be `max-width`?

File: custom/icds_reports/static/icds_reports/mobile/css/app.less [hunk 3]
```diff
@@ -326,3 +329,40 @@
...
+  max-height: 350px;
+  overflow: auto;
+  z-index: 1;
```
Comment (czue) [inline custom/icds_reports/static/icds_reports/mobile/css/app.less]: can this be moved to `varaibles.less`? we try to track all z-indices there so it's clear what should sit on top of what

File: custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+var url = hqImport('hqwebapp/js/initial_page_data').reverse;
+
+
+function SortableKpiController($rootScope, $scope) {
+    this.expandedCard = -1;
```
Comment (czue) [inline custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js]: nit: might be more readable if you had `const = NO_CARD_EXPANDED = -1`. or alternatively, why use `-1` and not `null` or `undefined`?

File: custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+var url = hqImport('hqwebapp/js/initial_page_data').reverse;
+
+
+function SortableKpiController($rootScope, $scope) {
```
Comment (czue) [inline custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js]: what is `$rootScope` used for?

File: custom/icds_reports/templates/icds_reports/icds_app/mobile/sortable-kpi.directive.html [hunk 6]
```diff
@@ -0,0 +1,26 @@
+{% load staticfiles %}
+<div ng-repeat="section in $ctrl.data">
+  <div class="shadow rounded-border bg-white col-xs-11 mx-auto float-unset mb-3 px-0 pb-3">
+    <div ng-click="$ctrl.toggleCardExpansion($index)">
+      <div class="d-flex justify-content-between align-items-center p-3" ng-click="$ctrl.expandCard($index)">
```
Comment (czue) [inline custom/icds_reports/templates/icds_reports/icds_app/mobile/sortable-kpi.directive.html]: where is `expandCard` defined?

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 7]
```diff
@@ -248,19 +253,22 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
...
-            return '<div> No expected ' + indicator + ' </div>';
+            return isMobile ? ('No expected ' + indicator) : '<div> No expected ' + indicator + ' </div>';
         }
         else {
             if (denominator === vm.dataNotEntered) { return vm.dataNotEntered; }
             if (percentage === vm.dataNotEntered) {
                 if (nominator === 0 && denominator === 0) {
-                    return '<div><span>100 %<br>(' + nominator + ' / ' + denominator + ')</span></div>';
+                    return isMobile ? ('100% (' + nominator + ' / ' + denominator + ')') :
+                        '<div><span>100 %<br>(' + nominator + ' / ' + denominator + ')</span></div>';
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: could this be extracted to a helper function? does all it do is add <div> and <span> tags to the web version?

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 8]
```diff
@@ -248,19 +253,22 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
...
         if (isZeroNullUnassignedOrDataNotEntered(denominator)) {
-            return '<div> No expected ' + indicator + ' </div>';
+            return isMobile ? ('No expected ' + indicator) : '<div> No expected ' + indicator + ' </div>';
         }
         else {
             if (denominator === vm.dataNotEntered) { return vm.dataNotEntered; }
             if (percentage === vm.dataNotEntered) {
                 if (nominator === 0 && denominator === 0) {
-                    return '<div><span>100 %<br>(' + nominator + ' / ' + denominator + ')</span></div>';
+                    return isMobile ? ('100% (' + nominator + ' / ' + denominator + ')') :
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: nit: not your fault but nominator --> numerator.

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 9]
```diff
@@ -337,17 +345,123 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
         }
     };
 
+    // mobile helpers
+    vm.showSortPopup = false;
+    vm.tooltipDisplayed = -1; // '-1' when none of the tooltips are displayed in sort popup
+    vm.requestDataStartingFrom = 0; // could be any multiple of 10
+    vm.dataSortingDirection = 0; // '0' -> asc, '1' -> desc
+    vm.sortingColumn = 0;
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: not sure if worth it but a lot of these could be made constants to make the code more readable. also i think a comment explaining that they map to the datatables API params might be helpful

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 10]
```diff
@@ -337,17 +345,123 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
...
+    vm.sortableInputKpiData = [];
+    vm.showSortingInfo = function (tooltipNumber) {
+        // when clicked on 'i' icon, get the column number on which info is requested and show only corresponding info.
+        vm.tooltipDisplayed = tooltipNumber;
+    };
+    vm.toggleSortPopup = function (event) {
+        vm.tooltipDisplayed = -1;
+        vm.showSortPopup = !vm.showSortPopup;
+        event.stopPropagation();
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: out of curiosity why is this needed?

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 11]
```diff
@@ -337,17 +345,123 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
...
+        vm.sortingColumn = 0;
+        vm.sortableInputKpiData = [];
+        vm.requestDataStartingFrom = 0;
+        vm.getData();
+        vm.toggleSortPopup(event);
+    };
+    vm.getMobileData = function (index) {
+        // triggers when clicked on any of the headings in sort popup
+        vm.dataSortingDirection = 1 - vm.dataSortingDirection;
+        vm.sortingColumn = index + 1;
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: not sure i understand this. why does the column and direction change? wouldn't direction only change if the column is the same?

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 12]
```diff
@@ -337,17 +345,123 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
...
+            case '2' :
+                vm.dataAggregationLevel = 4;
+                break;
+            case '3' :
+                vm.dataAggregationLevel = 5;
+                break;
+            default :
+                vm.dataAggregationLevel = 1;
+        }
+        var mobileCustomParams = {
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: think there should be a comment here about how this is spoofing the datatables API

**Review by czue — COMMENTED** [review]
This looks great - well done :clap: :clap: 

I had mostly nits and questions. Holding off on approval until you get a chance to review/respond

File: custom/icds_reports/static/icds_reports/mobile/css/app.less (hunk 2 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/icds_reports/mobile/css/app.less]: I have checked it with bigger headings, but it didnot overflow.. So I left it that way. But I will change it to max-width anyway. Thanks for pointing out

File: custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js (hunk 4 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js]: Yeah It can be null or undefined. I should have used strict comparision "===" rather than "==". where null and 0 are considered same.

I will make the change

File: custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js (hunk 5 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/sortable-kpi/sortable-kpi.directive.js]: I guess your question is "what is $rootScope used for in this component?"
I copied it from some other directive, and I forgot to remove it (both $rootscope and $scope)
I will remove it now

File: custom/icds_reports/templates/icds_reports/icds_app/mobile/sortable-kpi.directive.html (hunk 6 above)
Comment (saiprakash-v) [reply custom/icds_reports/templates/icds_reports/icds_app/mobile/sortable-kpi.directive.html]: I have changed the function name to `toggleCardExpansion` and moved it to the `div` above it. but didnot remove it.

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js (hunk 7 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: Sorry, I didn't get what you meant by extracting to helper function. which part?

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js (hunk 8 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: That totally went over my head

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js (hunk 10 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: At the top level element I have added a click event, which when triggered closes sort popup. But we dont need the click event which triggers the popup opening to bubble up to the top, which will close it. So, this prevents that click event from bubbling up the dom

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js (hunk 11 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: Correct. I will update the logic

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js [hunk 13]
```diff
@@ -346,36 +346,44 @@ function ServiceDeliveryDashboardController($rootScope, $scope, $http, $location
     };
 
     // mobile helpers
+    const NO_TOOLTIP_DISPLAYED = -1;
+    const DEFAULT_SORTED_COLUMN = 0;
+    const DEFAULT_SORTING_DIRECTION = 0;
```
Comment (czue) [inline custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: sorry i should have been a little more specific. one constant I had in mind, but didn't mention explicitly was having 
`SORT_ASCENDING = 0` and `SORT_DESCENDING = 1` which would improve readability. though is not important and maybe you can do it in a follow up so we can get this in first if you agree

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js (hunk 10 above)
Comment (czue) [reply custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: got it. that feels comment-worthy to me

**Review by czue — APPROVED** [review]
thanks, looks great!

File: custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js (hunk 7 above)
Comment (czue) [reply custom/icds_reports/static/js/directives/service-delivery-dashboard/service-delivery-dashboard.directive.js]: something like:

```javascript
function getDisplay(indicatorHtml) {
  return isMobile ? indicatorHtml : '<div><span>' + indicatorHtml + '</span></div>';
}
```

feel free to do in a follow up if i read it right and it sounds worthwhile. not a big deal, just a lot of repetition in that `indicatorHtml` bit currently

**Review by czue — APPROVED** [review]
